modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -331,6 +441,83 @@ public void onFailure(Exception e) {
         });
     }
 
+    /*
+     * This method executes the given force merge request. Once the request has completed successfully it writes a timestamp as custom
+     * metadata in the cluster state indicating when the force merge has completed. The listener is notified after the cluster state
+     * update has been made, or when the forcemerge fails or the write of the to the cluster state fails.
+     */
+    private void forceMergeIndex(ForceMergeRequest forceMergeRequest, ActionListener<Void> listener) {
+        assert forceMergeRequest.indices() != null && forceMergeRequest.indices().length == 1 : "DLM force merges one index at a time";
+        final String targetIndex = forceMergeRequest.indices()[0];
+        logger.info("DLM is issuing a request to force merge index [{}]", targetIndex);
+        client.admin().indices().forceMerge(forceMergeRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(ForceMergeResponse forceMergeResponse) {
+                if (forceMergeResponse.getFailedShards() > 0) {
+                    DefaultShardOperationFailedException[] failures = forceMergeResponse.getShardFailures();
+                    String message = Strings.format(
+                        "DLM failed to forcemerge %d shards for index [%s] due to failures [%s]",
+                        forceMergeResponse.getFailedShards(),
+                        targetIndex,
+                        failures == null
+                            ? "unknown"
+                            : Arrays.stream(failures).map(DefaultShardOperationFailedException::toString).collect(Collectors.joining(","))
+                    );
+                    logger.warn(message);
+                    listener.onFailure(new ElasticsearchException(message));
```

Comment:
I think we could replace this with:

```suggestion
                    this.onFailure(new ElasticsearchException(message));
```

And it will do the logging only once, instead of (potentially) spamming with the warning multiple times.

Reply:
I don't see the listener doing any logging at all -- it's going to `ErrorRecordingActionListener` which records the error, but I don't see it actually sending it to a logger, right? I also don't see multiple log messages coming out of `DataLifecycleServiceTests::testForceMergeRetries`.

Reply:
Looking at the code:

```java
            public void onFailure(Exception e) {
                String previousError = errorStore.getError(targetIndex);
                // To avoid spamming our logs, we only want to log the error once.
                if (previousError == null || previousError.equals(errorStore.getError(targetIndex)) == false) {
                    logger.warn(
                        () -> Strings.format(
                            "DLM encountered an error trying to force merge index [%s]. DLM will attempt to force merge the index on its "
                                + "next run.",
                            targetIndex
                        ),
                        e
                    );
                }
                listener.onFailure(e);
```

It will log it as a warning if the `previousError` is null or if it isn't the same

Reply:
> I don't see the listener doing any logging at all

Note that my version is calling `this.onFailure(...)` not `listener.onFailure(...)`, so it invokes the `onFailure` that does the only-once logging

Reply:
Oh! I had missed that you were changing which onFailure was being called! :)